utils/s3/client.cc

```diff
@@ -81,20 +82,30 @@ future<> ignore_reply(const http::reply& rep, input_stream<char>&& in_) {
     co_await util::skip_entire_stream(in);
 }
 
-static future<> look_for_errors(const http::reply&, input_stream<char>&& in_) {
-    auto in = std::move(in_);
-    auto body = co_await util::read_entire_stream_contiguous(in);
-    auto possible_error = aws::aws_error::parse(std::move(body));
-    if (possible_error && possible_error->get_error_type() != aws::aws_error_type::OK) {
-        throw std::system_error(std::error_code(EIO, std::generic_category()), possible_error->get_error_message());
+future<aws::aws_error> client::look_for_error(const http::reply& reply, input_stream<char>& in_, multipart_upload_completion is_mpu_completion) {
+    auto status_class = http::reply::classify_status(reply._status);
+    bool is_failure = status_class != http::reply::status_class::informational && status_class != http::reply::status_class::success;
+    if (is_failure || is_mpu_completion == multipart_upload_completion::yes) {
```

Comment:
What if few days after we find out that some other S3 query may return status_ok, then think and report actual error in the response body? Will we add `bool_class<struct some_other_query_that_behaves_like_that_tag>` and check it here too?

Please, make it look like

```
// generic retryable request maker
future<> s3::client::make_retryable_request(..., error_handler) {
   do {
        http_client->make_request();
        error = error_handler();
        if (!error) {
            return error;
        }
   } while (need_retry());
}

// retryable request maker with "default" error check, i.e. -- just check http status code
future<> s3::client::make_retryable_request(...) {
    return make_retryable_request(..., [] {
        return status == expected ? ok : error;
    });
}

// most of request don't need to do anything but check http status code
future<> client::pretty_much_any_operation() {
    return make_retryable_request(...);
}

// complete multipart upload needs some more sophisticated error checking, namely -- error may come in response body
future<> client::complete_multipart_upload() {
    return make_retryable_request(..., [] {
        if (status != ok)
            return error;
        if (error = co_await read_and_parse_body())
            return error;
        return ok;
    });
}
```

Reply:
We will not find out anything here. This is the only case so far in AWS S3 response that returns ok but carries possible error in the body. If AWS will introduce such additional case it will be super-breaking change not only for `awscli`/`aws_sdk`s but for many other applications around the world. I doubt they will do such a thing. But in any case, we can rewrite the code to be more generic.

In addition, you cant just
> retryable request maker with "default" error check, i.e. -- just check http status code

As we already discussed it [here](https://github.com/scylladb/scylladb/pull/21054#discussion_r1799254004), if the status code is not OK-ish one have to parse the body. In the MPU completion request we must always parse the body. So it should look more like

```
// generic retryable request maker
future<> s3::client::make_retryable_request(..., error_handler) {
   do {
        http_client->make_request();
        error = error_handler();
        if (!error) {
            return error;
        }
   } while (need_retry());
}

// retryable request maker with "default" error check
future<> s3::client::make_retryable_request(...) {
    return make_retryable_request(..., [] {
        if (status == expected)
            return ok;
        if (status == erroneous)
        {
            error = co_await read_and_parse_body();
            if (error)
                return error;
            return error_from_status(status);
        }
    });
}

// most of request don't need to do anything but check http status code
future<> client::pretty_much_any_operation() {
    return make_retryable_request(...);
}

// complete multipart upload needs some more sophisticated error checking, namely -- error may come in response body
future<> client::complete_multipart_upload() {
    return make_retryable_request(..., [] {
        error = co_await read_and_parse_body();
        if (error)
            return error;
        if (status == erroneous)
            return error_from_status(status);
        if (status == expected)
            return ok;
        return unexpected_status;
    });
}


```